### PR TITLE
Change MXNetTrainer to Estimator and remove RDD from plasma after training finishes

### PR DESCRIPTION
**Pull request by hkvision**

### CODE REVIEW

**Comment by jason-dai** [comment]
I think you should automatically clean the data in `MXNetTrainer.train`, not `shutdown`

**Comment by hkvision** [comment]
@jason-dai I have changed the API to Estimator.fit, data is specify in fit and would be removed after fit finishes. In theory, fit can be called multiple times for different data. Take a look.

File: pyzoo/zoo/orca/learn/mxnet/mxnet_trainer.py [hunk 1]
```diff
@@ -167,19 +115,77 @@ def __init__(self, config, train_data, model_creator,
...
-                test_data_list[i],
-                self.train_resize_batch_num,
                 self.eval_metrics_creator)
             for i, runner in enumerate(self.runners)
         ])
 
-    def train(self, nb_epoch=1):
-        """Trains an MXNet model for several epochs."""
-        stats = ray.get([w.train.remote(nb_epoch) for w in self.runners])
+    def train(self, train_data, val_data=None, nb_epoch=1, train_resize_batch_num=None):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/mxnet/mxnet_trainer.py]: fit

File: pyzoo/zoo/orca/data/shard.py [hunk 2]
```diff
@@ -139,8 +139,20 @@ def get_data(self):
...
+        if self.object_store_address:
+            if "client" not in self.__dict__:
+                import pyarrow.plasma as plasma
+                self.client = plasma.connect(self.object_store_address)
+            print("Removing data")
+            if self.client.contains(self.shard_list):
+                self.client.delete([self.shard_list])
+            assert not self.client.contains(self.shard_list)
+            print("Deleted successfully")
+            self.client.disconnect()
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/shard.py:155]: need to remove `client` from `self.__dict__`?

File: pyzoo/zoo/orca/data/shard.py [hunk 3]
```diff
@@ -139,8 +139,20 @@ def get_data(self):
         # The ObjectID would contain a list of data.
         else:
             import pyarrow.plasma as plasma
-            client = plasma.connect(self.object_store_address)
-            return client.get(self.shard_list)
+            self.client = plasma.connect(self.object_store_address)
+            return self.client.get(self.shard_list)
+
+    def __del__(self):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/shard.py:145]: need to catch possible errors in connect, delete, and disconnect?

File: pyzoo/zoo/orca/learn/mxnet/mxnet_trainer.py [hunk 4]
```diff
@@ -167,19 +115,77 @@ def __init__(self, config, train_data, model_creator,
...
+                val_data = val_data.repartition(self.num_workers).to_ray()
+        if isinstance(train_data, RayXShards):
+            if train_data.num_partitions() != self.num_workers:
+                train_data.repartition(self.num_workers)
+            if val_data:
+                assert isinstance(val_data, RayXShards)
+                if val_data.num_partitions() != self.num_workers:
+                    val_data.repartition(self.num_workers)
+
+        if isinstance(train_data, RayXShards):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/learn/mxnet/mxnet_trainer.py:165]: same condition as above?

File: pyzoo/zoo/orca/data/shard.py (hunk 2 above)
Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:155]: okay!

File: pyzoo/zoo/orca/data/shard.py (hunk 3 above)
Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:145]: What to do if there are errors? Just do nothing and print a warning?

File: pyzoo/zoo/orca/learn/mxnet/mxnet_trainer.py (hunk 4 above)
Comment (hkvision) [reply pyzoo/zoo/orca/learn/mxnet/mxnet_trainer.py:165]: yeah, would combine them into a single if-else.

File: pyzoo/zoo/orca/data/shard.py (hunk 3 above)
Comment (jason-dai) [reply pyzoo/zoo/orca/data/shard.py:145]: Maybe just report warning

Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:146]: Updated. Take a look?

Comment (jason-dai) [reply pyzoo/zoo/orca/data/shard.py:146]: Maybe just
```
try:
   if self.object_store_address:
      ...
except Exception as e:
   logger.warning(e)
   logger.warning("Error deleting the data from plasma")
```

Comment (hkvision) [reply pyzoo/zoo/orca/data/shard.py:146]: okay

**Comment by hkvision** [comment]
Tested RBI code on yarn for the modifications. Removing data from plasma after the training seems to be working properly.
Would merge this PR first.
If there are further comments, will address in a following PR.

**Comment by jason-dai** [comment]
LGTM

**Comment by jason-dai** [comment]
LGTM

**Comment by hkvision** [comment]
http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Validation/3663/

**Pull request merged**